crates/sui-graphql-rpc/src/types/epoch.rs

```diff
@@ -12,52 +12,131 @@ use super::transaction_block::{TransactionBlock, TransactionBlockFilter};
 use super::validator_set::ValidatorSet;
 use async_graphql::connection::Connection;
 use async_graphql::*;
+use sui_indexer::models_v2::epoch::StoredEpochInfo;
 
-#[derive(Clone, Debug, PartialEq, Eq, SimpleObject)]
-#[graphql(complex)]
+#[derive(Clone, Debug)]
 pub(crate) struct Epoch {
-    /// The epoch's id as a sequence number that starts at 0 and it is incremented by one at every epoch change
-    pub epoch_id: u64,
+    pub stored_epoch_info: StoredEpochInfo,
+}
+
+#[Object]
+impl Epoch {
     /// The epoch's protocol version
     #[graphql(skip)]
-    pub protocol_version: u64,
+    pub fn protocol_version(&self) -> u64 {
+        self.stored_epoch_info.protocol_version as u64
+    }
+
+    /// The epoch's id as a sequence number that starts at 0 and is incremented by one at every epoch change
+    async fn epoch_id(&self) -> u64 {
+        self.stored_epoch_info.epoch as u64
+    }
+
     /// The minimum gas price that a quorum of validators are guaranteed to sign a transaction for
-    pub reference_gas_price: Option<BigInt>,
+    async fn reference_gas_price(&self) -> Option<BigInt> {
+        Some(BigInt::from(
+            self.stored_epoch_info.reference_gas_price as u64,
+        ))
+    }
+
     /// Validator related properties, including the active validators
-    pub validator_set: Option<ValidatorSet>,
+    async fn validator_set(&self) -> Result<Option<ValidatorSet>> {
+        let validators = self
+            .stored_epoch_info
+            .validators
+            .clone()
+            .into_iter()
+            .flatten()
+            .map(|v| {
+                bcs::from_bytes(&v).map_err(|e| {
+                    Error::Internal(format!(
+                        "Can't convert validator into Validator. Error: {e}",
+                    ))
+                })
+            })
+            .collect::<Result<Vec<_>, Error>>()?;
+
+        let active_validators = convert_to_validators(validators, None);
+        let validator_set = ValidatorSet {
+            total_stake: self
+                .stored_epoch_info
+                .new_total_stake
+                .map(|s| BigInt::from(s as u64)),
+            active_validators: Some(active_validators),
+            ..Default::default()
+        };
+        Ok(Some(validator_set))
+    }
+
     /// The epoch's starting timestamp
-    pub start_timestamp: Option<DateTime>,
+    async fn start_timestamp(&self) -> Option<DateTime> {
+        DateTime::from_ms(self.stored_epoch_info.epoch_start_timestamp)
+    }
     /// The epoch's ending timestamp
-    pub end_timestamp: Option<DateTime>,
+    async fn end_timestamp(&self) -> Option<DateTime> {
+        self.stored_epoch_info
+            .epoch_end_timestamp
+            .and_then(DateTime::from_ms)
+    }
     /// The total number of checkpoints in this epoch.
-    pub total_checkpoints: Option<BigInt>,
+    async fn total_checkpoints(&self) -> Option<BigInt> {
```

Comment:
If we don't have the last checkpoint ID, it means that we are in the latest epoch, right? In which case we should look up the ID of the latest available checkpoint and use that, rather than returning `None` here.